File: fetch-tpex/scripts/fetch_tpex.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import datetime as dt
import json
import os
import sys
from typing import Any, Dict, List, Optional

import requests

TPEX_URL = "https://www.tpex.org.tw/web/stock/aftertrading/daily_close_quotes/stk_quote_result.php"


def to_roc(date_yyyymmdd: str) -> str:
    d = dt.datetime.strptime(date_yyyymmdd, "%Y%m%d").date()
    roc_year = d.year - 1911
    return f"{roc_year:03d}/{d.month:02d}/{d.day:02d}"


def parse_float(s: Any) -> Optional[float]:
    if s is None:
        return None
    s = str(s).strip().replace(",", "")
    if s in ("", "--", "-", "N/A"):
        return None
    # TPEX change can include trailing spaces
    try:
        return float(s)
    except Exception:
        return None


def parse_int(s: Any) -> Optional[int]:
    if s is None:
        return None
    s = str(s).strip().replace(",", "")
    if s in ("", "--", "-", "N/A"):
        return None
    try:
        return int(float(s))
    except Exception:
        return None


def fetch_table(date_yyyymmdd: str, timeout: int = 30) -> Dict[str, Any]:
    roc = to_roc(date_yyyymmdd)
    params = {"l": "zh-tw", "d": roc, "s": "0,asc,0", "o": "json"}
    headers = {"User-Agent": "Mozilla/5.0"}
    r = requests.get(TPEX_URL, params=params, headers=headers, timeout=timeout)
    r.raise_for_status()
    return r.json()


def extract_prices(j: Dict[str, Any], codes: List[str], date_yyyymmdd: str) -> Dict[str, Any]:
    stat = j.get("stat")
    if str(stat).upper() != "OK": 
        return {
            "source": "tpex",
            "fetchTime": dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
            "date": date_yyyymmdd,
            "dateROC": to_roc(date_yyyymmdd),
            "stocks": [],
            "error": {"type": "non-trading", "message": "TPEX stat not OK", "details": str(stat)},
        }

    tables = j.get("tables") or []
    if not tables or not isinstance(tables, list) or not isinstance(tables[0], dict):
        return {
            "source": "tpex",
            "fetchTime": dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
            "date": date_yyyymmdd,
            "dateROC": to_roc(date_yyyymmdd),
            "stocks": [],
            "error": {"type": "parse", "message": "Unexpected TPEX tables format", "details": None},
        }

    t0 = tables[0]
    fields = t0.get("fields") or []
    data = t0.get("data") or []

    # indices by field names
    def idx(name: str) -> int:
        try:
            return fields.index(name)
        except ValueError:
            return -1

    i_code = idx("代號")
    i_name = idx("名稱")
    i_close = idx("收盤")
    i_change = idx("漲跌")
    i_open = idx("開盤")
    i_high = idx("最高")
    i_low = idx("最低")
    i_vol = idx("成交股數")

    if min(i_code, i_name, i_open, i_close) < 0:
        return {
            "source": "tpex",
            "fetchTime": dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
            "date": date_yyyymmdd,
            "dateROC": to_roc(date_yyyymmdd),
            "stocks": [],
            "error": {"type": "parse", "message": "Missing expected fields", "details": fields},
        }

    wanted = set(codes)
    found: Dict[str, Dict[str, Any]] = {}

    for row in data:
        if not isinstance(row, list) or len(row) <= max(i_code, i_low, i_high, i_close, i_open):
            continue
        code = str(row[i_code]).strip()
        if code not in wanted:
            continue
        open_ = parse_float(row[i_open])
        close_ = parse_float(row[i_close])
        high_ = parse_float(row[i_high])
        low_ = parse_float(row[i_low])
        chg = parse_float(row[i_change]) if i_change >= 0 else None
        vol = parse_int(row[i_vol]) if i_vol >= 0 else None

        chg_pct = None
        if open_ is not None and close_ is not None and open_ != 0:
            chg_pct = (close_ - open_) / open_ * 100

        found[code] = {
            "code": code,
            "name": str(row[i_name]).strip(),
            "open": open_,
            "high": high_,
            "low": low_,
            "close": close_,
            "change": chg,
            "changePercent": None if chg_pct is None else round(chg_pct, 2),
            "volume": vol,
        }

    missing = [c for c in codes if c not in found]

    out = {
        "source": "tpex",
        "fetchTime": dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
        "date": date_yyyymmdd,
        "dateROC": to_roc(date_yyyymmdd),
        "stocks": [found.get(c) for c in codes],
        "error": None,
    }

    if missing:
        out["error"] = {
            "type": "not-found",
            "message": "Some codes not found in TPEX table",
            "details": missing,
        }

    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("date", help="YYYYMMDD")
    ap.add_argument("codes", nargs="+", help="stock codes (e.g., 6499 6610)")
    ap.add_argument("--json", action="store_true", help="output JSON only")
    args = ap.parse_args()

    try:
        j = fetch_table(args.date)
        out = extract_prices(j, args.codes, args.date)
    except Exception as e:
        out = {
            "source": "tpex",
            "fetchTime": dt.datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
            "date": args.date,
            "dateROC": to_roc(args.date),
            "stocks": [],
            "error": {"type": "exception", "message": str(e), "details": repr(e)},
        }

    if args.json:
        print(json.dumps(out, ensure_ascii=False))
        return

    # pretty
    print(json.dumps(out, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: fetch-tpex/scripts/fetch_tpex.mjs
import axios from 'axios';
import fs from 'fs';
import path from 'path';

/**
 * TPEX (櫃買中心) 股價抓取程式
 * 目的：抓取上櫃個股收盤資料 (全市場)
 * 依賴：axios
 * 
 * 用法:
 * node fetch_tpex.mjs [stockCode] [outputPath]
 * 
 * 參數:
 * 1. stockCode (選填): 指定股票代號 (例如: "6499") 或 "all" (預設)。若要篩選多檔，請在代碼間用逗號分隔 (例如: "6499,6610")。
 * 2. outputPath (選填): 儲存結果的檔案路徑 (例如: /path/to/tpex.json)。若未提供，則根據日期與代碼自動生成檔名。
 * 
 * 範例:
 * node fetch_tpex.mjs all ./data/tpex_20260210.json
 * node fetch_tpex.mjs 6499
 */

const args = process.argv.slice(2);
const stockCodeArg = args[0] || 'all'; // Arg 1: stockCode or 'all'
const outputPath = args[1]; // Arg 2: outputPath

// 取得今日日期 (YYYYMMDD)
const dateStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');

// 解析目標代碼
let targetCodes = [];
if (stockCodeArg.toLowerCase() !== 'all') {
    targetCodes = stockCodeArg.split(',');
}

// 轉換西元 -> 民國 (YYYYMMDD -> YYY/MM/DD)
function toRocDate(yyyymmdd) {
    const year = parseInt(yyyymmdd.substring(0, 4)) - 1911;
    const month = yyyymmdd.substring(4, 6);
    const day = yyyymmdd.substring(6, 8);
    return `${year}/${month}/${day}`;
}

async function fetchTpex() {
    try {
        const rocDate = toRocDate(dateStr);
        const url = `https://www.tpex.org.tw/web/stock/aftertrading/daily_close_quotes/stk_quote_result.php?l=zh-tw&d=${rocDate}&s=0,asc,0&o=json`;

        console.log(`Fetching TPEX data: ${dateStr} (${rocDate})`);
        console.log(`Target: ${stockCodeArg === 'all' ? 'All Market' : targetCodes.join(', ')}`);
        console.log(`URL: ${url}`);

        const response = await axios.get(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            },
            timeout: 15000
        });

        const data = response.data;

        // TPEX 回傳的 aaData 為主要資料
        if (!data.aaData || data.aaData.length === 0) {
             console.error('TPEX API Error: No data found (aaData is empty)');
             if (data.reportDate !== rocDate) {
                 console.log('Possibly a holiday or no data available.');
             }
             return;
        }

        let resultData = data.aaData;

        // 若有指定代碼，進行篩選 (TPEX API 不支援直接指定代碼，需 client-side filter)
        if (targetCodes.length > 0) {
            resultData = resultData.filter(row => targetCodes.includes(row[0])); // row[0] 是代碼
        }

        const output = {
            source: 'tpex',
            date: dateStr,
            count: resultData.length,
            data: resultData
        };

        // 輸出 JSON 到 stdout
        const jsonOutput = JSON.stringify(output, null, 2);
        console.log('JSON_OUTPUT_START');
        console.log(jsonOutput);
        console.log('JSON_OUTPUT_END');

        // 決定儲存路徑
        let filename;
        if (outputPath) {
            filename = outputPath;
            const dir = path.dirname(filename);
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
        } else {
            // 預設檔名
            filename = targetCodes.length > 0
                ? `tpex_${targetCodes.join('_')}_${dateStr}.json`
                : `tpex_${dateStr}.json`;
        }

        fs.writeFileSync(filename, jsonOutput, 'utf-8');
        console.log(`Saved to ${filename}`);

    } catch (error) {
        console.error('Error fetching TPEX data:', error.message);
    }
}

fetchTpex();
